src/fmc.rs

```diff
@@ -0,0 +1,278 @@
+
+// From stm32_fmc
+use stm32_fmc::FmcPeripheral;
+use stm32_fmc::{PinsSdram, Sdram, SdramChip, SdramPinSet, SdramTargetBank};
+
+use crate::rcc::{rec, rec::ResetEnable, CoreClocks};
+use crate::stm32;
+use crate::time::Hertz;
+
+use crate::gpio::gpioa::PA7;
+use crate::gpio::gpiob::{PB5, PB6, PB7};
+use crate::gpio::gpioc::{PC0, PC2, PC3, PC4, PC5, PC6, PC7, PC8};
+use crate::gpio::gpiod::{
+    PD0, PD1, PD10, PD11, PD12, PD13, PD14, PD15, PD3, PD4, PD5, PD6, PD7, PD8,
+    PD9,
+};
+use crate::gpio::gpioe::{
+    PE0, PE1, PE10, PE11, PE12, PE13, PE14, PE15, PE2, PE3, PE4, PE5, PE6, PE7,
+    PE8, PE9,
+};
+use crate::gpio::gpiof::{
+    PF0, PF1, PF11, PF12, PF13, PF14, PF15, PF2, PF3, PF4, PF5,
+};
+use crate::gpio::gpiog::{
+    PG0, PG1, PG10, PG12, PG13, PG14, PG15, PG2, PG3, PG4, PG5, PG6, PG7, PG8,
+    PG9,
+};
+use crate::gpio::gpioh::{
+    PH10, PH11, PH12, PH13, PH14, PH15, PH2, PH3, PH5, PH6, PH7, PH8, PH9,
+};
+use crate::gpio::gpioi::{PI0, PI1, PI10, PI2, PI3, PI4, PI5, PI6, PI7, PI9};
+use crate::gpio::{Alternate, AF12, AF9};
+
+/// Storage type for Flexible Memory Controller and its clocks
+pub struct FMC {
+    fmc: stm32::FMC,
+    _prec: rec::Fmc,
+    fmc_ker_ck: Option<Hertz>,
+}
+
+/// Extension trait for FMC controller
+pub trait FmcExt: Sized {
+    fn fmc(self, prec: rec::Fmc, clocks: &CoreClocks) -> FMC;
+
+    /// A new SDRAM memory via the Flexible Memory Controller
+    fn sdram<BANK: SdramPinSet, PINS: PinsSdram<BANK>, CHIP: SdramChip>(
+        self,
+        pins: PINS,
+        chip: CHIP,
+        prec: rec::Fmc,
+        clocks: &CoreClocks,
+    ) -> Sdram<FMC, CHIP> {
+        let fmc = self.fmc(prec, clocks);
+        Sdram::new(fmc, pins, chip)
+    }
+
+    /// A new SDRAM memory via the Flexible Memory Controller
+    fn sdram_unchecked<CHIP: SdramChip, BANK: Into<SdramTargetBank>>(
+        self,
+        bank: BANK,
+        chip: CHIP,
+        prec: rec::Fmc,
+        clocks: &CoreClocks,
+    ) -> Sdram<FMC, CHIP> {
+        let fmc = self.fmc(prec, clocks);
+        Sdram::new_unchecked(fmc, bank, chip)
+    }
+}
+
+impl FmcExt for stm32::FMC {
+    /// New FMC instance
+    fn fmc(self, prec: rec::Fmc, clocks: &CoreClocks) -> FMC {
+        let clk_sel = prec.get_kernel_clk_mux();
+
+        // Calculate kernel clock
+        let fmc_ker_ck = match clk_sel {
+            rec::FmcClkSel::RCC_HCLK3 => Some(clocks.hclk()),
+            rec::FmcClkSel::PLL1_Q => clocks.pll1_q_ck(),
+            rec::FmcClkSel::PLL2_R => clocks.pll2_r_ck(),
+            rec::FmcClkSel::PER => clocks.per_ck(),
+        };
+
+        FMC {
+            fmc: self,
+            _prec: prec,
+            fmc_ker_ck,
+        }
+    }
+}
+
+unsafe impl Sync for FMC {}
+unsafe impl FmcPeripheral for FMC {
+    const REGISTERS: *const () = stm32::FMC::ptr() as *const ();
+
+    fn enable(&mut self) {
+        // unsafe: we own this via the PREC structure
+        let rcc = unsafe { &*stm32::RCC::ptr() };
```

Comment:
I'm not understanding why we have to unsafely take access to the RCC. Wasn't the PREC structure designed to implement the necessary functions below? E.g. enable the peripheral and reset it?

Reply:
So the methods in [`ResetEnable`](https://docs.rs/stm32h7xx-hal/0.6.0/stm32h7xx_hal/rcc/rec/trait.ResetEnable.html) require us to own the PREC structure (`self`). However the `FmcPeripheral` only gives us a mutable reference (`&mut self`) so we can't call them.

This highlights that the PREC `reset`/`enable` methods are only usable at state transitions when we own the PREC alone, or we are constructing / destructing a struct that contains it.

Perhaps the problem is in the `FmcPeripheral` trait? One option would be to remove the `enable` method entirely. Instead we would document that implementing the `FmcPeripheral` trait requires the AHB bus clock to be enabled already (and we would use PREC to enable it in `FmcExt::new`).

Reply:
As a workaround to these restrictions, I think you could wrap the `ResetEnable` (PREC) in an [`Option`](https://doc.rust-lang.org/core/option/enum.Option.html). You can then [`take()`](https://doc.rust-lang.org/core/option/enum.Option.html#method.take) it during the enable section to have full ownership of the PREC and then return it into the option at the end.

However, I wanted to comment that this feels like more of a workaround to the fact that `ResetEnable` doesn't take a mutable ref. What's the reasoning behind passing ownership of the `ResetEnable` and consuming `self` instead of just acceptable a mutable ref?

Reply:
Agreed this is a good point - I made a separate issue to track this - https://github.com/stm32-rs/stm32h7xx-hal/issues/117